### DISCUSSION TITLE
wasm package size

### DISCUSSION THREAD

**Discussion by yiwenxue** [cocos/cocos-engine#15944] [General]
## Meshoptimizer decoder wasm package

[meshoptimizer repo](https://github.com/zeux/meshoptimizer)

| filesize        | emscripten (with stl) | emscripten | emscripten (-s) | wasi                  |
| --------------- | --------------------- | ---------- | --------------- | --------------------- |
| decoder.wasm.js | 102 KB                | 95.8 KB    | 95.8 KB         | 0 KB + (6 KB helper) |
| decoder.wasm    | 320 KB                | 243 KB     | 20.0 KB         | 5 KB                  |
| decoder.asm.js  | 369 KB                | 243 KB     | 244 KB          | 47 KB                 |
| sum             | 791 KB                | 581.8 KB   | 359 KB          | 58 KB                 |

emscripten (-s):

```bash
D:/meshlet/emsdk/upstream/emscripten/em++.bat  @CMakeFiles/decoder.dir/includes_CXX.rsp -g -O3 -DNDEBUG -nostartfiles -nostdlib -Wl,--no-entry -fno-slp-vectorize -fno-vectorize -fno-unroll-loops -MD -MT CMakeFiles/decoder.dir/external/meshoptimizer/src/vertexcodec.cpp.o -MF CMakeFiles/decoder.dir/external/meshoptimizer/src/vertexcodec.cpp.o.d -o CMakeFiles/decoder.dir/external/meshoptimizer/src/vertexcodec.cpp.o -c D:/meshlet/meshopt/external/meshoptimizer/src/vertexcodec.cpp

D:/meshlet/emsdk/upstream/emscripten/em++.bat  @CMakeFiles/decoder.dir/includes_CXX.rsp -g -O3 -DNDEBUG -nostartfiles -nostdlib -Wl,--no-entry -fno-slp-vectorize -fno-vectorize -fno-unroll-loops -MD -MT CMakeFiles/decoder.dir/external/meshoptimizer/src/vertexcodec.cpp.o -MF CMakeFiles/decoder.dir/external/meshoptimizer/src/vertexcodec.cpp.o.d -o CMakeFiles/decoder.dir/external/meshoptimizer/src/vertexcodec.cpp.o -c D:/meshlet/meshopt/external/meshoptimizer/src/vertexcodec.cpp

D:/meshlet/emsdk/upstream/emscripten/em++.bat -g --bind --allow-undefined -Wl,-s -Wl,--no-entry -s TOTAL_MEMORY=128MB --stack-size=32768 -s NO_EXIT_RUNTIME=1 -s ASSERTIONS=2 -s WASM=1 -s MODULARIZE=1 -s ALLOW_MEMORY_GROWTH=1 -s EXPORTED_FUNCTIONS="['_malloc','_free','_meshopt_decodeIndexBuffer','_meshopt_decodeVertexBuffer']" -s EXPORTED_RUNTIME_METHODS="['ccall', 'cwrap']" @CMakeFiles/decoder.dir/objects1.rsp -o decoder.wasm.js
```

wasi sdk:

```bash
/d/workspace/bullet-compile/extenal/wasi-sdk-20.0/bin/clang++ src/vertexcodec.cpp src/indexcodec.cpp src/vertexfilter.cpp tools/wasmstubs.cpp --target=wasm32-wasi --sysroot=/d/workspace/bullet-compile/extenal/wasi-sdk-20.0/share/wasi-sysroot -O3 -DNDEBUG -nostartfiles -nostdlib -Wl,--no-entry -Wl,-s -fno-slp-vectorize -fno-vectorize -fno-unroll-loops -Wl,-z -Wl,stack-size=24576 -Wl,--initial-memory=65536 -Wl,--export=meshopt_decodeVertexBuffer -Wl,--export=meshopt_decodeIndexBuffer -Wl,--export=sbrk -Wl,--export=__wasm_call_ctors -o build/decoder_base.wasm
```

**Comment by yiwenxue** [comment]
emscripten cmake project: 
```cmake
  set (DECODER_WASM_SOURCES
  ${MESHOPT_DIR}/src/indexcodec.cpp
  ${MESHOPT_DIR}                                                                           
  /src/vertexcodec.cpp
    ${MESHOPT_DIR}/tools/wasmstubs.cpp
  )

  set (DECODER_WASM_EXPORTS
    # allocator
    _malloc
    _free

    # decoder
    _meshopt_decodeIndexBuffer
    _meshopt_decodeVertexBuffer
  )

  # add prefix to exports
  foreach (WASM_EXPORT ${DECODER_WASM_EXPORTS})
    set (WASM_EXPORT "'${WASM_EXPORT}'")
    list (APPEND DECODE_EXPORTS ${WASM_EXPORT})
  endforeach ()

  string (REPLACE ";" "," DECODE_EXPORTS "${DECODE_EXPORTS}")

  set (CXX_OPTIONS "-O3 -DNDEBUG -nostartfiles -nostdlib -fno-vectorize -fno-unroll-loops ")
  set (LINK_OPTIONS  "--bind --allow-undefined -Wl,-s -Wl,--no-entry -s TOTAL_MEMORY=128MB --stack-size=32768 -s NO_EXIT_RUNTIME=1 -s ASSERTIONS=2 -s WASM=0 -s ALLOW_MEMORY_GROWTH=1 -s EXPORTED_FUNCTIONS=\"[${DECODE_EXPORTS}]\" -s EXPORTED_RUNTIME_METHODS=\"['ccall', 'cwrap']\"")

  add_executable(decoder ${DECODER_WASM_SOURCES})

  target_include_directories(decoder
    PRIVATE
      ${CMAKE_SOURCE_DIR}/src
      ${MESHOPT_DIR}/src
    )

  set_target_properties(decoder PROPERTIES COMPILE_FLAGS ${CXX_OPTIONS})
  set_target_properties(decoder PROPERTIES LINK_FLAGS ${LINK_OPTIONS})
  set_target_properties(decoder PROPERTIES OUTPUT_NAME "decoder.wasm")
```

**Comment by yiwenxue** [comment]
[compare.zip](https://github.com/cocos/cocos-engine/files/12332095/compare.zip)

- emscripten:
  - decoder.wasm.js
  - decoder.wasm
  - decoder.asm.js
- wasi:
  - meshopt_decoder.wasm.js (decoder.wasm.js + decoder.wasm + decoder.wasm(simd version))
  - meshopt_decoder.asm.js (generated by wasm2js)

**Comment by dumganhar** [comment]
LINK_OPTIONS also needs to add `-O3` flag

**Comment by yiwenxue** [comment]
| filesize        | ems1   | ems2     | ems3    | ems4     | wasi                 |
| --------------- | ------ | -------- | ------- | -------- | -------------------- |
| decoder.wasm.js | 102 KB | 95.8 KB  | 95.8 KB | 48.9 KB  | 0 KB + (6 KB helper) |
| decoder.wasm    | 320 KB | 243 KB   | 20.0 KB | 18.0 KB  | 5 KB                 |
| decoder.asm.js  | 369 KB | 243 KB   | 244 KB  | 84.4 KB  | 47 KB                |
| sum             | 791 KB | 581.8 KB | 359 KB  | 151.3 KB | 58 KB                |

ems1:

compile_flags:  "-O3 -DNDEBUG"

link_flags: "-g --bind --allow-undefined -Wl,--no-entry -s TOTAL_MEMORY=128MB --stack-size=32768 -s NO_EXIT_RUNTIME=1 -s ASSERTIONS=2 -s WASM=0 -s ALLOW_MEMORY_GROWTH=1 -s EXPORTED_FUNCTIONS=\"[${DECODE_EXPORTS}]\" -s EXPORTED_RUNTIME_METHODS=\"['ccall', 'cwrap']\""

ems2:

compile_flags:  "-O3 -DNDEBUG ***-nostartfiles -nostdlib -fno-vectorize -fno-unroll-loops***"

link_flags: "-g --bind --allow-undefined -Wl,--no-entry -s TOTAL_MEMORY=128MB --stack-size=32768 -s NO_EXIT_RUNTIME=1 -s ASSERTIONS=2 -s WASM=0 -s ALLOW_MEMORY_GROWTH=1 -s EXPORTED_FUNCTIONS=\"[${DECODE_EXPORTS}]\" -s EXPORTED_RUNTIME_METHODS=\"['ccall', 'cwrap']\""

ems3:

compile_flags:  "-O3 -DNDEBUG -nostartfiles -nostdlib -fno-vectorize -fno-unroll-loops"

link_flags: "-g --bind --allow-undefined ***-Wl,-s*** -Wl,--no-entry -s TOTAL_MEMORY=128MB --stack-size=32768 -s NO_EXIT_RUNTIME=1 -s ASSERTIONS=2 -s WASM=0 -s ALLOW_MEMORY_GROWTH=1 -s EXPORTED_FUNCTIONS=\"[${DECODE_EXPORTS}]\" -s EXPORTED_RUNTIME_METHODS=\"['ccall', 'cwrap']\""

ems4 (release):

compile_flags: "-O3 -DNDEBUG -nostartfiles -nostdlib -fno-vectorize -fno-unroll-loops ***-fno-exceptions -fno-rtti***"

link_flags: "***~~-g~~ -O3*** --bind --allow-undefined -Wl,-s -Wl,--no-entry -s TOTAL_MEMORY=128MB --stack-size=32768 -s NO_EXIT_RUNTIME=1 -s ASSERTIONS=2 -s WASM=1 -s MODULARIZE=1 -s ALLOW_MEMORY_GROWTH=1 -s EXPORTED_FUNCTIONS=\"[${DECODE_EXPORTS}]\" -s EXPORTED_RUNTIME_METHODS=\"['ccall', 'cwrap']\"


[compress2.zip](https://github.com/cocos/cocos-engine/files/12333168/compress2.zip)

- emscripten:
  - decoder.wasm.js
  - decoder.wasm
  - decoder.asm.js
  - decoder.asm.js.mem
- wasi:
  - meshopt_decoder.wasm.js (decoder.wasm.js + decoder.wasm + decoder.wasm(simd version))
  - meshopt_decoder.asm.js (generated by wasm2js)

**Comment by yiwenxue** [comment]
| filesize        | ems1   | ems2     | ems3    | ems4     | ems5    | wasi                 |
| --------------- | ------ | -------- | ------- | -------- | ------- | -------------------- |
| decoder.wasm.js | 102 KB | 95.8 KB  | 95.8 KB | 48.9 KB  | 30.5 KB | 0 KB + (6 KB helper) |
| decoder.wasm    | 320 KB | 243 KB   | 20.0 KB | 18.0 KB  | 13.2 KB | 5 KB                 |
| decoder.asm.js  | 369 KB | 243 KB   | 244 KB  | 84.4 KB  | 58.2 KB | 47 KB                |
| sum             | 791 KB | 581.8 KB | 359 KB  | 151.3 KB | 102 KB  | 58 KB                |
## Flags
### ems1:

compile_flags:  "-O3 -DNDEBUG"

link_flags: "-g --bind --allow-undefined -Wl,--no-entry -s TOTAL_MEMORY=128MB --stack-size=32768 -s NO_EXIT_RUNTIME=1 -s ASSERTIONS=2 -s WASM=0 -s ALLOW_MEMORY_GROWTH=1 -s EXPORTED_FUNCTIONS=\"[${DECODE_EXPORTS}]\" -s EXPORTED_RUNTIME_METHODS=\"['ccall', 'cwrap']\""

### ems2:

compile_flags:  "-O3 -DNDEBUG ***-nostartfiles -nostdlib -fno-vectorize -fno-unroll-loops***"

link_flags: "-g --bind --allow-undefined -Wl,--no-entry -s TOTAL_MEMORY=128MB --stack-size=32768 -s NO_EXIT_RUNTIME=1 -s ASSERTIONS=2 -s WASM=0 -s ALLOW_MEMORY_GROWTH=1 -s EXPORTED_FUNCTIONS=\"[${DECODE_EXPORTS}]\" -s EXPORTED_RUNTIME_METHODS=\"['ccall', 'cwrap']\""

### ems3:

compile_flags:  "-O3 -DNDEBUG -nostartfiles -nostdlib -fno-vectorize -fno-unroll-loops"

link_flags: "-g --bind --allow-undefined ***-Wl,-s*** -Wl,--no-entry -s TOTAL_MEMORY=128MB --stack-size=32768 -s NO_EXIT_RUNTIME=1 -s ASSERTIONS=2 -s WASM=0 -s ALLOW_MEMORY_GROWTH=1 -s EXPORTED_FUNCTIONS=\"[${DECODE_EXPORTS}]\" -s EXPORTED_RUNTIME_METHODS=\"['ccall', 'cwrap']\""

### ems4 (release):

compile_flags: "-O3 -DNDEBUG -nostartfiles -nostdlib -fno-vectorize -fno-unroll-loops ***-fno-exceptions -fno-rtti***"

link_flags: "***~~-g~~ -O3*** --bind --allow-undefined -Wl,-s -Wl,--no-entry -s TOTAL_MEMORY=128MB --stack-size=32768 -s NO_EXIT_RUNTIME=1 -s ASSERTIONS=2 -s WASM=1 -s MODULARIZE=1 -s ALLOW_MEMORY_GROWTH=1 -s EXPORTED_FUNCTIONS=\"[${DECODE_EXPORTS}]\" -s EXPORTED_RUNTIME_METHODS=\"['ccall', 'cwrap']\"

### ems5 (release):

compile_flags: "-O3 -DNDEBUG -nostartfiles -nostdlib -fno-vectorize -fno-unroll-loops -fno-exceptions -fno-rtti"

link_flags: "-O3 ***~~--bind --allow-undefined~~*** ***-s FILESYSTEM=0 -s ERROR_ON_UNDEFINED_SYMBOLS=0 -s DYNAMIC_EXECUTION=0*** -Wl,-s -Wl,--no-entry -s TOTAL_MEMORY=128MB --stack-size=32768 -s NO_EXIT_RUNTIME=1 -s ASSERTIONS=2 -s WASM=1 -s MODULARIZE=1 -s ALLOW_MEMORY_GROWTH=1 -s EXPORTED_FUNCTIONS=\"[${DECODE_EXPORTS}]\" -s EXPORTED_RUNTIME_METHODS=\"['ccall', 'cwrap']\"

## summary
1. release (no -g)
   strip symbols, strip wasm.js
	wasm.js  -= 45KB
	asm.js -= 162 KB
1. -nostartfiles -nostdlib
   total -= 210 KB
3. -fno-exceptions -fno-rtti
    wasm -= 3KB
4. linker -O3 -Wl,-s (strip all symbol)
   strip symbols even with -g
   wasm -= 223 KB
5. -s FILESYSTEM=0 -s ERROR_ON_UNDEFINED_SYMBOLS=0 -s DYNAMIC_EXECUTION=0
   wasm.js -= 1 KB
6. --bind --allow-undefined
   wasm.js -= 19 KB
   wasm -= 5 KB
   asm.js -= 26 KB
